Feature/parse only

rachmadaniHaryono opened the pull request:
~~WIP~~

- add `print-only` option to print direct link from an album.
- add new extractor method using json library, see note below
- merge test using pytest parametrize
- fix args variable
- test json parsing
- add click dependancy
- readme explanation: new feature and xargs wget example
- test on all format for this json method.
- fix wrong cli command on readme

not yet commited
- ~`missing image test` on regex method (optional)~ skipped, use only the available imgur urls on test file. 

i test it using this imgur album http://imgur.com/a/stfnm (nsfw) and got different result between new method with json and old method with regex. maybe there is a way to double check the album using both method.

opinion @jtara1?

e: update after succesful test

rachmadaniHaryono committed:
chg: dev: Use click to parse the argument

rachmadaniHaryono committed:
new: dev: add click as required package

rachmadaniHaryono committed:
new: usr: print only feature

jtara1 commented on imgurdownloader/imgurdownloader.py:
```diff
@@ -135,29 +138,66 @@ def __init__(self, imgur_url, dir_download=os.getcwd(), file_name='',
+            try:
+                search = search.group().replace('\n', '', ).split(':', 1)[1].rsplit('}', 1)[0]
+                json_search = json.loads(search)
+                img_dicts = json_search['album_images']['images']
```
So, you're converting part of the html to dict via json instead of using regex to parse it? That's fine by me, but it looks like you're only considering album imgur links?

e.g.: view-source:http://imgur.com/FVRUGe2

This e.g. would fail to match any ids with your json conversion and search I believe.

rachmadaniHaryono replied:
it would fail but json search will still a valid json

here is what i got from json search 

```yaml
{'adConfig': {'safeFlags': ['onsfw_mod_safe', 'share', 'page_load'], 'showsAds': True, 'unsafeFlags': []},
 'animated': False,
 'bandwidth': 0,
 'date': '9 months ago',
 'datetime': '2016-09-15 00:12:30',
 'deletehash': None,
 'description': '',
 'ext': '.png',
 'favorited': False,
 'hash': 'FVRUGe2',
 'height': '1200',
 'isAd': False,
 'is_moderated': False,
 'is_safe': True,
 'is_viral': 0,
 'looping': False,
 'name': None,
 'prefer_video': False,
 'size': 2085906,
 'source': '',
 'title': 'Lock and Load [Stella Jogakuin Koukouka C3-bu]',
 'video_host': None,
 'video_source': None,
 'views': '2487',
 'width': '835'}
```
i will fix it to include single image format.

rachmadaniHaryono committed:
fix: dev: json search on single image format

jtara1 commented:
What's the reason for using click over the built in lib, argparse https://docs.python.org/3/library/argparse.html

rachmadaniHaryono commented:
simple and clearer arg command(see below) and more utils (which may or may not needed) and better subcommand.

for usual argparse in most of the program i read is located on single program. example:

```python
def parse_args(argv):
    parser = argparse.ArgumentParser(description='Process some integers.')
    parser.add_argument('integers', metavar='N', type=int, nargs='+',
                    help='an integer for the accumulator')
    parser.add_argument('--sum', dest='accumulate', action='store_const',
                    const=sum, default=max,
                    help='sum the integers (default: find the max)')
    # ... another argument here
    return parser.parse_args(argv)

# after this function is called the actual function is called and the args 
# actual main 
def main():
    args = parse_args(sys.argv[1:])
    command(args)    
```
for me it is kinda redundant, and click can possibly short that all out.

```python
@click.command()
@click.option(...)
@click.argument(...)
def command(...):
    # function here
```

jtara1 commented:
I added the list of all imgur urls I expect this lib to support to the test file https://github.com/jtara1/imgur_downloader/blob/master/tests/test_image_ids_and_extensions.py

Feel free to add some test functions there.

If your json_search returns a nice dict with all that info like that for all of those urls then we could refactor a bunch of the code to just use that.

rachmadaniHaryono commented:
can i refactor that test? it look like it can use pytest parametrize

jtara1 commented:
yea

rachmadaniHaryono committed:
chg: test: merge similar test using pytest parametrize !refactor

jtara1 commented:
Ready to merge with master?

rachmadaniHaryono commented:
Not yet. Maybe I can't complete it today (my laptop got error and I response it from my tablet) 

I will let you know if it is complete and cross re WIP sign on first post

rachmadaniHaryono committed:
fix: dev: flake8 on setup.py

rachmadaniHaryono committed:
fix: dev: flake8 and args variable on imgurdownloader

rachmadaniHaryono committed:
new: test: test json parsing

- test both method at the same time
- remove unused library

rachmadaniHaryono committed:
chg: dev: refactor json parse

- set the init result directly to class attr
- copy if it is direct image
- splitted error
- use yield even if only one image found

rachmadaniHaryono commented:
done with the test already, i will update the first post based on change from the last comment.

travis check result

https://travis-ci.org/rachmadaniHaryono/imgur_downloader/builds/245297071

for the readme section, i will update it tommorow

rachmadaniHaryono committed:
new: doc: `print only` feature on readme

rachmadaniHaryono commented:
i think it is done for now. i may replace or give the user choice to which parser they choose and change the default to json parser on next pr. right now this is enough for `print-only` option.
